render.c

```diff
@@ -39,89 +105,45 @@ int render(struct mako_state *state, struct pool_buffer *buffer) {
-		pango_layout_set_wrap(layout, PANGO_WRAP_WORD_CHAR);
-		pango_layout_set_ellipsize(layout, PANGO_ELLIPSIZE_END);
-		PangoFontDescription *desc =
-			pango_font_description_from_string(config->font);
-		pango_layout_set_font_description(layout, desc);
-		pango_font_description_free(desc);
 
 		size_t text_len = format_notification(notif, config->format, NULL);
 		char *text = malloc(text_len + 1);
 		if (text == NULL) {
 			break;
 		}
 		format_notification(notif, config->format, text);
-
-		if (config->markup) {
-			PangoAttrList *attrs = NULL;
-			char *buf = NULL;
-			GError *error = NULL;
-			if (!pango_parse_markup(text, -1, 0, &attrs, &buf, NULL, &error)) {
-				fprintf(stderr, "cannot parse pango markup: %s\n",
-					error->message);
-				g_error_free(error);
-				break; // TODO: better error handling
-			}
-			pango_layout_set_markup(layout, buf, -1);
-			pango_layout_set_attributes(layout, attrs);
-			pango_attr_list_unref(attrs);
-			free(buf);
-		} else {
-			pango_layout_set_text(layout, text, -1);
-		}
-		free(text);
-
-		int text_height = 0;
-		pango_layout_get_pixel_size(layout, NULL, &text_height);
-		int notif_height = border_size + padding_size + text_height;
-
+		int notif_y = height;
 		if (i > 0) {
-			height += inner_margin;
+			notif_y += inner_margin;
 		}
-		int notif_y = height;
-		height += notif_height;
-
-		// Render border
-		set_cairo_source_u32(cairo, config->colors.border);
-		cairo_set_line_width(cairo, border_size);
-		cairo_rectangle(cairo, 0, notif_y, notif_width, notif_height);
-		cairo_stroke(cairo);
-
-		// Render background
-		set_cairo_source_u32(cairo, config->colors.background);
-		cairo_set_line_width(cairo, 0);
-		cairo_rectangle(cairo,
-			config->border_size, notif_y + config->border_size,
-			notif_width - border_size, notif_height - border_size);
-		cairo_fill(cairo);
-
-		// Render text
-		set_cairo_source_u32(cairo, config->colors.text);
-		cairo_move_to(cairo, config->border_size + config->padding,
-			notif_y + config->border_size + config->padding);
-		pango_cairo_update_layout(cairo, layout);
-		pango_cairo_show_layout(cairo, layout);
+
+		int notif_height = render_notification(cairo, state, text, notif_y);
+		height = notif_y +  notif_height;
 
 		// Update hotspot
 		notif->hotspot.x = 0;
 		notif->hotspot.y = notif_y;
 		notif->hotspot.width = notif_width;
 		notif->hotspot.height = notif_height;
 
-		g_object_unref(layout);
-
 		++i;
 		if (config->max_visible >= 0 &&
 				i >= (size_t)config->max_visible) {
 			break;
 		}
 	}
 
+	if (wl_list_length(&state->notifications) > config->max_visible) {
+		int hidden = wl_list_length(&state->notifications) - config->max_visible;
+		int hidden_ln = snprintf(NULL, 0, "[%d]", hidden);
+
+		char *hidden_text;
+		hidden_text = malloc(hidden_ln + 1);
+		snprintf(hidden_text, hidden_ln + 1, "[%d]", hidden);
+
+		int hidden_height = render_notification(cairo, state, hidden_text, height);
```

Comment:
We need to add `inner_margin` to `height` before this line, otherwise the hidden count notification collides with the previous one.

Reply:
The reason for this was that personally I liked the look of the count being closer to the notifications than the spacing provided by `inner_margin` but I added your change.

Reply:
Ah, I see. In the future, this can probably be added back as a config option somehow.